
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="BrainDataPortal is a platform for brian multi-modal data exploration and visualization.">
      
      
        <meta name="author" content="Ruifeng Hu">
      
      
        <link rel="canonical" href="https://huruifeng.github.io/BrainDataPortal/prepare_dataset/xqtl/">
      
      
        <link rel="prev" href="../visiumst/">
      
      
        <link rel="next" href="../../development/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>xQTL dataset - BrainDataPortal Docs</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.84d31ad4.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/css/style.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#xqtl-dataset-preparation" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="BrainDataPortal Docs" class="md-header__button md-logo" aria-label="BrainDataPortal Docs" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            BrainDataPortal Docs
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              xQTL dataset
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/huruifeng/BrainDataPortal" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    huruifeng/BrainDataPortal
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="BrainDataPortal Docs" class="md-nav__button md-logo" aria-label="BrainDataPortal Docs" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    BrainDataPortal Docs
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/huruifeng/BrainDataPortal" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    huruifeng/BrainDataPortal
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../install/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    Install
    
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Install
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
          
        
      
        
      
        
      
        
      
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    Prepare dataset
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3" id="__nav_3_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Prepare dataset
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scrnaseq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    scRNAseq dataset
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visiumst/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    VisiumST dataset
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    xQTL dataset
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    xQTL dataset
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-prerequisites" class="md-nav__link">
    <span class="md-ellipsis">
      1. Prerequisites
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-download-demo-data" class="md-nav__link">
    <span class="md-ellipsis">
      2. Download demo data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      3. Preprocessing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-mapping-cell-type-names-to-files" class="md-nav__link">
    <span class="md-ellipsis">
      4. Mapping cell type names to files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5global-significance-filtering" class="md-nav__link">
    <span class="md-ellipsis">
      5.Global significance filtering
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-generate-annotation-jsons" class="md-nav__link">
    <span class="md-ellipsis">
      6. Generate annotation JSONs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-convert-to-parquet" class="md-nav__link">
    <span class="md-ellipsis">
      7. Convert to Parquet
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-clean-up-unnecessary-files" class="md-nav__link">
    <span class="md-ellipsis">
      8. Clean up unnecessary files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#9-prepare-gwas-data-for-visualization" class="md-nav__link">
    <span class="md-ellipsis">
      9. Prepare GWAS data for visualization
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#10-prepare-dataset-information-file" class="md-nav__link">
    <span class="md-ellipsis">
      10. Prepare dataset information file
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#11-upload-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      11. Upload dataset
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../development/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    Development
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4" id="__nav_4_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Development
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../development/customize_home/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Customize Home
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../development/add_view/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Add view
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../contribute/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    Contribute
    
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Contribute
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-prerequisites" class="md-nav__link">
    <span class="md-ellipsis">
      1. Prerequisites
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-download-demo-data" class="md-nav__link">
    <span class="md-ellipsis">
      2. Download demo data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      3. Preprocessing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-mapping-cell-type-names-to-files" class="md-nav__link">
    <span class="md-ellipsis">
      4. Mapping cell type names to files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5global-significance-filtering" class="md-nav__link">
    <span class="md-ellipsis">
      5.Global significance filtering
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-generate-annotation-jsons" class="md-nav__link">
    <span class="md-ellipsis">
      6. Generate annotation JSONs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-convert-to-parquet" class="md-nav__link">
    <span class="md-ellipsis">
      7. Convert to Parquet
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-clean-up-unnecessary-files" class="md-nav__link">
    <span class="md-ellipsis">
      8. Clean up unnecessary files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#9-prepare-gwas-data-for-visualization" class="md-nav__link">
    <span class="md-ellipsis">
      9. Prepare GWAS data for visualization
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#10-prepare-dataset-information-file" class="md-nav__link">
    <span class="md-ellipsis">
      10. Prepare dataset information file
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#11-upload-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      11. Upload dataset
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="xqtl-dataset-preparation">xQTL Dataset Preparation<a class="headerlink" href="#xqtl-dataset-preparation" title="Permanent link">&para;</a></h1>
<p>Step-by-step guide for preparing single-cell ATAC-seq/xQTL/GWAS data for visualization in BrainDataPortal.</p>
<p>Learn how to prepare and process single-cell/nuclei ATAC-seq/xQTL/GWAS data for visualization in BrainDataPortal. 
This section covers seurat object processing, gene expression data splitting, metadata table preparation and data formatting.</p>
<p>We will use a brain dataset as an example and cover all essential preprocessing steps.</p>
<h2 id="1-prerequisites">1. Prerequisites<a class="headerlink" href="#1-prerequisites" title="Permanent link">&para;</a></h2>
<ul>
<li>Python 3.8+ with pandas, numpy, json, polars, fastparquet, pyBigWig libraries installed.</li>
<li>Basic understanding of single-cell ATAC-seq/QTL concepts.</li>
</ul>
<h2 id="2-download-demo-data">2. Download demo data<a class="headerlink" href="#2-download-demo-data" title="Permanent link">&para;</a></h2>
<p>We will use a single-cell dataset from human brain. 
This demo dataset contains several cell types from human brain middle brain region.</p>
<ul>
<li>Demo dataset and scripts:<blockquote>
<ol>
<li>Seurat object: <a href="https://pd5d.yale.edu/demos/notebooks/xqtl/example_data/eQTL_Demo.zip">xQTL_Demo.zip</a></li>
<li>Gene location file: <a href="https://pd5d.yale.edu/demos/notebooks/xqtl/example_data/gene_annotations.tsv">gene_annotations.tsv</a></li>
<li>SNP location file: <a href="https://pd5d.yale.edu/demos/notebooks/xqtl/example_data/snp_annotations.tsv">snp_annotations.tsv</a></li>
<li>GWAS summary file: <a href="https://pd5d.yale.edu/demos/notebooks/xqtl/example_data/PD_GWAS_with_rsIDs_hg38.tab">PD_GWAS.tab</a></li>
<li>Dataset configuration file: <a href="../../demos/notebooks/xqtl/dataset_info.toml">dataset_info.toml</a></li>
<li>Processing script: <a href="../../demos/scripts/xqtl_script.zip">xqtl_script.zip</a></li>
</ol>
</blockquote>
</li>
</ul>
<h2 id="3-preprocessing">3. Preprocessing<a class="headerlink" href="#3-preprocessing" title="Permanent link">&para;</a></h2>
<p>In this step, we will read the raw xQTL summary files and extract the necessary columns and filter the data based on the p-values.</p>
<ul>
<li>This script should vary depending on the input data.</li>
<li>Renames QTL data columns to `gene_id, snp_id, p_value, beta_value'.</li>
<li>Splits QTL data by celltype into <code>unfiltered_celltypes/</code>.</li>
<li>Renames gene and SNP annotation columns to <code>gene_id</code>,<code>position_start</code>, <code>position_end</code>, <code>strand</code> and <code>snp_id</code>, <code>position</code>.</li>
<li>Filters each celltype in <code>unfiltered_celltypes/</code> for entries with p &gt; 0.01 and outputs filtered TSVs into <code>filtered_celltypes/</code>.</li>
</ul>
<p>Full code in Python: <a href="../../demos/scripts/xqtl/01_preprocessing.py">01_preprocessing.py</a>.</p>
<p>Modify the following parameters according to your dataset and run the script:
<div class="highlight"><pre><span></span><code><span class="o">...</span> <span class="o">...</span>
<span class="c1">############################</span>
<span class="c1">## define parameters, modify as needed</span>
<span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;eQTLsummary_demo&quot;</span>
<span class="n">qtl_data_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;eQTLsummary_sampled/*_sampled.tsv&quot;</span><span class="p">))</span>
<span class="n">geneid_col</span> <span class="o">=</span> <span class="s2">&quot;gene_id&quot;</span>
<span class="n">snpid_col</span> <span class="o">=</span> <span class="s2">&quot;variant_id&quot;</span>
<span class="n">pvalue_col</span> <span class="o">=</span> <span class="s2">&quot;pval_nominal&quot;</span>
<span class="n">beta_col</span> <span class="o">=</span> <span class="s2">&quot;slope&quot;</span>
<span class="o">...</span> <span class="o">...</span>
</code></pre></div></p>
<div class="admonition warning">
<p class="admonition-title">Important Note</p>
<p>The QTL summary files should be in the TSV format with the following columns:</p>
<p><h4> Example xQTL file format (For each cell type,e.g. Astrocytes_eQTL.tsv):</h4></p>
<div class="highlight"><pre><span></span><code>Gene    SNP p-value beta
A1BG    rs1234567   0.02    0.213
A1BG    rs1234568   0.03    0.314
A1BG    rs1234569   0.01    0.615
</code></pre></div>
</div>
<h2 id="4-mapping-cell-type-names-to-files">4. Mapping cell type names to files<a class="headerlink" href="#4-mapping-cell-type-names-to-files" title="Permanent link">&para;</a></h2>
<p>In this step, we will map the cell type names to the files. 
It prompts user for display names that correspond to filenames in<code>celltypes/</code>. 
Otherwise the <strong>file names</strong> will be used.</p>
<p>The python script is: <a href="../../demos/scripts/xqtl/02_celltype_mapping.py">02_celltype_mapping.py</a>.</p>
<div class="admonition warning">
<p class="admonition-title">Important Note</p>
<p>Change the <strong>dataset_name = "your_dataset_name"</strong> to your dataset name. </p>
</div>
<h2 id="5global-significance-filtering">5.Global significance filtering<a class="headerlink" href="#5global-significance-filtering" title="Permanent link">&para;</a></h2>
<p>This step identifies gene-SNP pairs significant in at least one celltype using <code>filtered_celltypes/</code>.
Filters <code>unfiltered_celltypes/</code> accordingly, and outputs filtered TSVs into <code>celltypes/</code>.</p>
<p>Full code in Python: <a href="../../demos/scripts/xqtl/03_significant.py">03_significant.py</a>.</p>
<h2 id="6-generate-annotation-jsons">6. Generate annotation JSONs<a class="headerlink" href="#6-generate-annotation-jsons" title="Permanent link">&para;</a></h2>
<p>This step generates annotation JSONs for genes and SNPs.
The script reads the <code>gene_annotations.tsv</code> and <code>snp_annotations.tsv</code> files and splits gene and SNP annotation files by chromosome into <code>gene_locations/</code> and <code>snp_locations/</code></p>
<p>Full code in Python: <a href="../../demos/scripts/xqtl/04_annotate.py">04_annotate.py</a>.</p>
<p>Set the following parameters according to your dataset:
<div class="highlight"><pre><span></span><code><span class="c1">## Dataset specific parameters</span>
<span class="c1">## define parameters, modify as needed</span>
<span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;eQTLsummary_demo&quot;</span>
<span class="n">gene_annotation_file</span> <span class="o">=</span> <span class="s2">&quot;data/gene_annotations.tsv&quot;</span>
<span class="n">geneid_col</span> <span class="o">=</span> <span class="s2">&quot;gene_id&quot;</span>
<span class="n">gene_start_col</span> <span class="o">=</span> <span class="s2">&quot;position_start&quot;</span>
<span class="n">gene_end_col</span> <span class="o">=</span> <span class="s2">&quot;position_end&quot;</span>
<span class="n">gene_chrom_col</span> <span class="o">=</span> <span class="s2">&quot;chromosome&quot;</span> 
<span class="n">gene_strand_col</span> <span class="o">=</span> <span class="s2">&quot;strand&quot;</span>  

<span class="n">snp_annotation_file</span> <span class="o">=</span> <span class="s2">&quot;data/snp_annotations.tsv&quot;</span>
<span class="n">snpid_col</span> <span class="o">=</span> <span class="s2">&quot;snp_id&quot;</span>
<span class="n">snp_chrom_col</span> <span class="o">=</span> <span class="s2">&quot;chr&quot;</span>
<span class="n">snp_pos_col</span> <span class="o">=</span> <span class="s2">&quot;position&quot;</span>
</code></pre></div></p>
<div class="admonition warning">
<p class="admonition-title">Important Note</p>
<p>You can use the demo annotation files provided in the demo data. Or you can prepare your own annotation files. 
The annotation files should be in the TSV format with the following columns:</p>
<p><h4> Gene annotation file format (TSV)</h4></p>
<div class="highlight"><pre><span></span><code>gene    chromosome  start   end strand
A1BG    1   1234567 1234568 -
A1BG    1   1234568 1234569 -
A1BG    1   1234569 1234570 -
</code></pre></div>
<p><h4> SNP annotation file format (TSV)</h4></p>
<div class="highlight"><pre><span></span><code>SNP chromosome  position
rs1234567   1   1234567
rs1234568   1   1234568
rs1234569   1   1234569
</code></pre></div>
</div>
<h2 id="7-convert-to-parquet">7. Convert to Parquet<a class="headerlink" href="#7-convert-to-parquet" title="Permanent link">&para;</a></h2>
<p>This step converts all necessary TSV files to Parquet format for final use.</p>
<p>Full code: <a href="../../demos/scripts/xqtl/05_parquet.py">05_parquet.py</a>.</p>
<p>Modify the following codes for your specific dataset.
<div class="highlight"><pre><span></span><code><span class="c1"># Define dataset name</span>
<span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;eQTLsummary_demo&quot;</span>
<span class="o">...</span> <span class="o">...</span>
</code></pre></div></p>
<h2 id="8-clean-up-unnecessary-files">8. Clean up unnecessary files<a class="headerlink" href="#8-clean-up-unnecessary-files" title="Permanent link">&para;</a></h2>
<p>This step cleans up unnecessary files. It removes the <code>unfiltered_celltypes/</code> directory and the <code>filtered_celltypes/</code> directory.</p>
<p>Full code: <a href="../../demos/scripts/xqtl/06_filter_clean.py">06_filter_clean.py</a>.</p>
<p>Modify the following codes for your specific dataset.
<div class="highlight"><pre><span></span><code><span class="c1"># Define dataset name</span>
<span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;eQTLsummary_demo&quot;</span>
<span class="o">...</span> <span class="o">...</span>
</code></pre></div></p>
<h2 id="9-prepare-gwas-data-for-visualization">9. Prepare GWAS data for visualization<a class="headerlink" href="#9-prepare-gwas-data-for-visualization" title="Permanent link">&para;</a></h2>
<p>BrainDataPortal can visualize GWAS summary data in the form of a scatter plot.
We provided a script to split the GWAS summary data to required TSV format files.</p>
<p>Full code: <a href="../../demos/scripts/xqtl/07_gwas.py">07_gwas.py</a>.</p>
<p>Modify the following codes for your specific dataset.
<div class="highlight"><pre><span></span><code><span class="c1">## Dataset specific parameters</span>
<span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;eQTLsummary_demo&quot;</span>
<span class="n">input_file</span> <span class="o">=</span> <span class="s2">&quot;PD_GWAS_with_rsIDs_hg38.tab&quot;</span>
<span class="n">gwas_chrom_col</span> <span class="o">=</span> <span class="s2">&quot;#chrom&quot;</span>
<span class="n">gwas_pos_col</span> <span class="o">=</span> <span class="s2">&quot;chromEnd&quot;</span>
<span class="n">gwas_snp_col</span> <span class="o">=</span> <span class="s2">&quot;rsID&quot;</span>
<span class="n">gwas_beta_col</span> <span class="o">=</span> <span class="s2">&quot;b&quot;</span>
<span class="n">gwas_pval_col</span> <span class="o">=</span> <span class="s2">&quot;p&quot;</span>
<span class="o">...</span> <span class="o">...</span>
</code></pre></div></p>
<h2 id="10-prepare-dataset-information-file">10. Prepare dataset information file<a class="headerlink" href="#10-prepare-dataset-information-file" title="Permanent link">&para;</a></h2>
<p>Download and modify the demo dataset information file: <a href="../../demos/notebooks/xqtl/dataset_info.toml">dataset_info.toml</a></p>
<div class="admonition danger">
<p class="admonition-title">Important Note</p>
<p>Dataset configuration file name must be <strong>dataset_info.toml</strong>.</p>
</div>
<p>Here is an example of the dataset configuration file content:
<div class="highlight"><pre><span></span><code><span class="k">[datasetfile]</span>
<span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                               </span><span class="c1">## Path to the Seurat object file</span>
<span class="n">datatype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                           </span><span class="c1">## Type of the data. Options: scRNAseq, scATACseq, VisiumST, xQTL</span>

<span class="k">[dataset]</span>
<span class="n">dataset_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                       </span><span class="c1">## Required: Dataset name, MUST BE UNIQUE, used to identify the dataset in the database</span>
<span class="n">description</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                        </span><span class="c1">## Dataset description</span>
<span class="n">PI_full_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                       </span><span class="c1">## Principal Investigator (PI) full name</span>
<span class="n">PI_email</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                           </span><span class="c1">## PI email</span>
<span class="n">first_contributor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                  </span><span class="c1">## First contributor name</span>
<span class="n">first_contributor_email</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">            </span><span class="c1">## First contributor email</span>
<span class="n">other_contributors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                 </span><span class="c1">## Other contributors</span>
<span class="n">support_grants</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                     </span><span class="c1">## Support grants</span>
<span class="n">other_funding_source</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">               </span><span class="c1">## Other funding source</span>
<span class="n">publication_DOI</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                    </span><span class="c1">## DOI of the publication</span>
<span class="n">publication_PMID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                   </span><span class="c1">## PMID of the publication</span>
<span class="n">brain_super_region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                 </span><span class="c1">## Brain super region</span>
<span class="n">brain_region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                       </span><span class="c1">## Brain region</span>
<span class="n">sample_info</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                        </span><span class="c1">## Sample information</span>
<span class="n">sample_sheet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                       </span><span class="c1">## Sample sheet file name (Not the path, just the file name)</span>
<span class="n">n_samples</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">96</span><span class="w">                          </span><span class="c1">## Number of samples</span>
<span class="n">organism</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;Homo Sapiens&quot;</span><span class="w">               </span><span class="c1">## Organism</span>
<span class="n">tissue</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;Brain&quot;</span><span class="w">                        </span><span class="c1">## Tissue</span>
<span class="n">disease</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;PD&quot;</span><span class="w">                          </span><span class="c1">## Disease</span>

<span class="k">[study]</span>
<span class="n">study_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;Parkinson5D&quot;</span><span class="w">              </span><span class="c1">## Study name, the dataset belongs to</span>
<span class="n">description</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                        </span><span class="c1">## Study description</span>
<span class="n">team_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;Team Scherzer&quot;</span><span class="w">             </span><span class="c1">## Team name</span>
<span class="n">lab_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;NeuroGenomics&quot;</span><span class="w">              </span><span class="c1">## Lab name</span>
<span class="n">submitter_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                     </span><span class="c1">## Submitter name</span>
<span class="n">submitter_email</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                    </span><span class="c1">## Submitter email</span>

<span class="k">[protocol]</span>
<span class="n">protocol_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;P002&quot;</span><span class="w">                    </span><span class="c1">## Protocol ID</span>
<span class="n">protocol_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;P001_VisiumST&quot;</span><span class="w">         </span><span class="c1">## Protocol name</span>
<span class="n">version</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                            </span><span class="c1">## Protocol version</span>
<span class="n">github_url</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                         </span><span class="c1">## GitHub URL</span>
<span class="n">sample_collection_summary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">          </span><span class="c1">## Sample collection summary</span>
<span class="n">cell_extraction_summary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">            </span><span class="c1">## Cell extraction summary</span>
<span class="n">lib_prep_summary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                   </span><span class="c1">## Library preparation summary</span>
<span class="n">data_processing_summary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">            </span><span class="c1">## Data processing summary</span>
<span class="n">protocols_io_DOI</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                   </span><span class="c1">## protocols.io DOI</span>
<span class="n">other_reference</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="w">                    </span><span class="c1">## Other reference</span>

<span class="k">[meta_features]</span>
<span class="n">selected_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span><span class="s2">&quot;nCount_Spatial&quot;</span><span class="p">,</span><span class="err">...] ## List of selected features will be shown in the page</span>
<span class="n">sample_id_column</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;sample_name&quot;</span><span class="w">        </span><span class="c1">## Sample ID column in Seurat object metadata</span>
<span class="n">major_cluster_column</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;CellType&quot;</span><span class="w">       </span><span class="c1">## Major cluster column in Seurat object metadata</span>
<span class="n">condition_column</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;diagnosis&quot;</span><span class="w">          </span><span class="c1">## Condition column in Seurat object metadata        </span>

<span class="k">[visium_defaults]</span>
<span class="n">samples</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span><span class="w"> </span><span class="s2">&quot;BN2023&quot;</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;BN1076&quot;</span><span class="p">,]</span><span class="w">         </span><span class="c1">## List of sample names</span>
<span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span><span class="w"> </span><span class="s2">&quot;smoothed_label_s5&quot;</span><span class="p">,</span><span class="err">...]    ## List of default feature names</span>
<span class="n">genes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span><span class="w"> </span><span class="s2">&quot;SNCA&quot;</span><span class="p">,</span><span class="err">...]                    ## List of default gene names</span>
</code></pre></div></p>
<h2 id="11-upload-dataset">11. Upload dataset<a class="headerlink" href="#11-upload-dataset" title="Permanent link">&para;</a></h2>
<p>Upload the dataset to server folder.</p>
<p>After running the pipeline, there will be a dataset folder that contains all necessary files:</p>
<ul>
<li>The files and folders within the <code>your_dataset_name</code> folder are necessary for upload.</li>
<li>Upload the dataset folder named '&lt;dataset_name&gt;' to the server at <strong>'backend/datasets'</strong>.</li>
<li>Put the dataset configuration file named <strong>'dataset_info.toml'</strong> to your dataset folder at <strong>'backend/datasets/&lt;dataset_name&gt;'</strong>.</li>
<li>Refresh the database: Go to <strong>'Datasets Management Page '(DATASETS -&gt; +ADD DATASET)</strong> and click <strong>'REFRESH DB'</strong>.</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
    
        <div class="md-copyright__highlight">
        Copyright &copy; 2024 Ruifeng Hu
        </div>
    
    
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
        Material for MkDocs
        </a>
        and localized using the
        <a href="https://ultrabug.github.io/mkdocs-static-i18n/" target="_blank" rel="noopener">
        MkDocs Static i18n Plugin
        </a>
    
    </div>
      
        
<div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/huruifeng" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://www.linkedin.com/in/ruifeng-hu-702b9383/" target="_blank" rel="noopener" title="www.linkedin.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3M135.4 416H69V202.2h66.5V416zM102.2 96a38.5 38.5 0 1 1 0 77 38.5 38.5 0 1 1 0-77m282.1 320h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9z"/></svg>
    </a>
  
    
    
    
    
    <a href="mailto:huruifeng.cn@hotmail.com" target="_blank" rel="noopener" title="" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M48 64C21.5 64 0 85.5 0 112c0 15.1 7.1 29.3 19.2 38.4l208 156a48 48 0 0 0 57.6 0l208-156c12.1-9.1 19.2-23.3 19.2-38.4 0-26.5-21.5-48-48-48zM0 196v188c0 35.3 28.7 64 64 64h384c35.3 0 64-28.7 64-64V196L313.6 344.8c-34.1 25.6-81.1 25.6-115.2 0z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.indexes", "navigation.instant.progress", "navigation.tracking", "navigation.sections", "navigation.path", "navigation.top"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../assets/js/myscript.js"></script>
      
    
  </body>
</html>